src/tum_dlr_automl_for_eo/utils/custom_metrics.py

```diff
@@ -1,4 +1,102 @@
 import numpy as np
+from random import sample
+
+
+def get_neighbors(architecture):
+    """
+    @param architecture: 1 binary encoded architecture np.array
+    @return: 5 neighbors of given arc. with 1 Hamming Distance
+    """
+    # get 5 random indices to create 5 neighbors
+    random_neighbor_indices = sample(range(0, len(architecture) - 1), 5)
+    neighbors = []
+    for index in random_neighbor_indices:
+        # Change architecture to keep Hamming Distance at 1
+        if architecture[index] == 0:
+            curr_architecture = architecture.copy()
+            curr_architecture[index] = 1
+        else:
+            curr_architecture = architecture.copy()
+            curr_architecture[index] = 0
+        neighbors.append(curr_architecture)
+    return neighbors
+
+
+def perform_bils_algorithm(architectures, acc_list, starting_point):
+    """
+
+    @param architectures: np.array [(Number of Models) x (length of binary encoded architecture)] binary encoded
+    architectures
+    @param acc_list: np.array [Number of Models] Final Accuracies of Models @param starting_point: int
+    Starting model to search
+    @return: list [found architecture index, found architecture's accuracy]
+             in case of no improvement returns given architecture info.
+    """
+    MAX_NUMBER_OF_ITERATIONS = 50
+
+    architectures = architectures
+    curr_architecture_index = starting_point
+
+    counter = 0
+
+    while counter < MAX_NUMBER_OF_ITERATIONS:
+
+        curr_architecture = architectures[curr_architecture_index]
+        current_acc = acc_list[curr_architecture_index]
+        found_arc_index = curr_architecture_index
+        found_acc = current_acc
+
+        neighbors = get_neighbors(curr_architecture)
+
+        # Loop over obtained 5 neighbors
+        for neighbor in neighbors:
+            for arc_index, arc in enumerate(architectures):
+
+                if all(neighbor == arc) and acc_list[arc_index] > current_acc:
+                    found_arc_index = arc_index
+                    found_acc = acc_list[found_arc_index]
+                    # print(
+                    #     f"from:{curr_architecture_index} to:{found_arc_index} pre_acc:{current_acc} found_acc:{found_acc}")
```

Comment:
I just wanted to ask the understand the BILS algorithm. In here, we compare the neighbours of an architecture with the all previous architectures and if the existing architecture has better accuracy we count that as a found architecture right?

Reply:
Yes exactly